MixtralSparseMoeBlock: add gate jitter

lorenzoverardo opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

This PR adds gate jitter to MixtralSparseMoeBlock's input data before passing it through the MoE layer, if turned on.
Fixes https://github.com/huggingface/transformers/issues/29429.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [X] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [X] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

- text models: @ArthurZucker and @younesbelkada

ArthurZucker commented on src/transformers/models/mixtral/configuration_mixtral.py:
```diff
@@ -93,6 +93,8 @@ class MixtralConfig(PretrainedConfig):
             allow the model to output the auxiliary loss. See [here]() for more details
         router_aux_loss_coef (`float`, *optional*, defaults to 0.001):
             The aux loss factor for the total loss.
+        router_jitter_noise (`float`, *optional*, defaults to 0.01):
```
thanks for following the API conventions! Let's default to 0 to make sur this is BC

lorenzoverardo replied:
I've defaulted it to `0.0` as it's a float, and to keep it consistent with the other args 🙂

ArthurZucker commented on src/transformers/models/mixtral/configuration_mixtral.py:
```diff
@@ -134,6 +136,7 @@ def __init__(
         num_local_experts=8,
         output_router_logits=False,
         router_aux_loss_coef=0.001,
+        router_jitter_noise=0.01,
```
```suggestion
        router_jitter_noise=0,
```

ArthurZucker commented on src/transformers/models/mixtral/modeling_mixtral.py:
```diff
@@ -837,9 +837,21 @@ def __init__(self, config):
+        self.jitter_noise = config.router_jitter_noise
+
+    def jitter(self, x):
+        if self.training and self.jitter_noise > 0:
+            low = 1.0 - self.jitter_noise
+            high = 1.0 + self.jitter_noise
+            noise = torch.rand(x.size(), dtype=x.dtype, device=x.device)
+            return x * (low + noise * (high - low))
+        return x
```
no need for a new function, let's do something like https://github.com/huggingface/transformers/blob/b345f4e7f7f852f7def0ebe473d21419ae235fdc/src/transformers/models/switch_transformers/modeling_switch_transformers.py#L163

ArthurZucker reviewed (commented):
Thanks!

ArthurZucker reviewed (approved):
Last nit and good to go!

lorenzoverardo committed:
MixtralSparseMoeBlock: add gate jitter

This commit adds gate jitter to MixtralSparseMoeBlock's input data
before passing it through the MoE layer, if turned on.

ArthurZucker commented:
Thanks for the contribution!

ArthurZucker commented:
FYI @younesbelkada

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_29865). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.
